### PR TITLE
Non blocking map load

### PR DESCRIPTION
**Pull request by RoiEXLab**
This PR does the map Parsing off the EDT when clicking "Select Map", so the GUI doesn't freeze.
Additionally a wating dialog is being displayed during that time, so the user knows the engine is actually doing something.

### CODE REVIEW

**Comment by RoiEXLab** [comment]
Just again for clarification: https://github.com/triplea-game/triplea/pull/2572/commits/583613daffbdeb3f29cb8ff85868afb9065b4aa4 was originally intended for #2573, but was moved here because of merge conflicts

File: src/main/java/games/strategy/engine/framework/startup/ui/GameSelectorPanel.java [hunk 1]
```diff
@@ -370,27 +372,53 @@ private void selectGameFile(final boolean saved) {
...
-      final NewGameChooserEntry entry =
-          NewGameChooser.chooseGame(JOptionPane.getFrameForComponent(this), model.getGameName());
-      if (entry != null) {
-        if (!entry.isGameDataLoaded()) {
+      new Thread(() -> {
+        final NewGameChooserEntry entry =
+            NewGameChooser.chooseGame(JOptionPane.getFrameForComponent(this), model.getGameName());
+        if (entry != null) {
+          final AtomicReference<WaitDialog> dialogReference = new AtomicReference<>();
+          SwingUtilities.invokeLater(() -> {
```
Comment (RoiEXLab) [inline src/main/java/games/strategy/engine/framework/startup/ui/GameSelectorPanel.java]: I'll use the BackgroudTaskRunner here...
Will change that

**Comment by RoiEXLab** [comment]
@ssoloff Turns out the BackgroundTaskRunner is actually pretty useful.
Made a lot of the complex code in this PR uneccessary.
This PR is now ready for review.

**Comment by RoiEXLab** [comment]
Unfortunately I had to drop the NewGameChooserModel test, simply because due to the changes to the code it now results in a waiting dialog as side-effect.

I'm not 100% satisfied with this solution, if you can tell me how to use a swing worker or similar, to have a blocking, but actually not blocking operation on the EDT, like when using BackgroundTaskRunner.runInBackgroundAndReturn simply without the dialog. I'd be happy to implement it this way...

**Comment by ssoloff** [comment]
> This PR is now ready for review.

I'm going offline soon, so I'll try to get to it sometime tomorrow.

> if you can tell me how to use a swing worker or similar, to have a blocking, but actually not blocking operation on the EDT, like when using BackgroundTaskRunner.runInBackgroundAndReturn simply without the dialog

You can use a [`SecondaryLoop`](https://docs.oracle.com/javase/8/docs/api/java/awt/SecondaryLoop.html) to simulate what a modal dialog does in terms of blocking the EDT at the point `setVisible(true)` is called while keeping UI events pumping.  If you were to modify `BackgroundTaskRunner#runInBackgroundAndReturnOrThrow()` to use a `SecondaryLoop` instead of a `WaitDialog`, it would look like:

```diff
diff --git a/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java b/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java
index 878a31894..fe6b0ed6d 100644
--- a/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java
+++ b/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java
@@ -3,6 +3,8 @@ package games.strategy.engine.framework.ui.background;
 import static com.google.common.base.Preconditions.checkNotNull;
 import static com.google.common.base.Preconditions.checkState;
 
+import java.awt.SecondaryLoop;
+import java.awt.Toolkit;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
@@ -83,7 +85,7 @@ public final class BackgroundTaskRunner {
 
     final AtomicReference<T> resultRef = new AtomicReference<>();
     final AtomicReference<Throwable> exceptionRef = new AtomicReference<>();
-    final WaitDialog waitDialog = GameRunner.newWaitDialog(message);
+    final SecondaryLoop loop = Toolkit.getDefaultToolkit().getSystemEventQueue().createSecondaryLoop();
     final SwingWorker<T, Void> worker = new SwingWorker<T, Void>() {
       @Override
       protected T doInBackground() throws Exception {
@@ -92,20 +94,19 @@ public final class BackgroundTaskRunner {
 
       @Override
       protected void done() {
-        waitDialog.setVisible(false);
-        waitDialog.dispose();
-
         try {
           resultRef.set(get());
         } catch (final ExecutionException e) {
           exceptionRef.set(e.getCause());
         } catch (final InterruptedException e) {
           exceptionRef.set(e);
+        } finally {
+          loop.exit();
         }
       }
     };
     worker.execute();
-    waitDialog.setVisible(true);
+    loop.enter();
 
     final @Nullable Throwable exception = exceptionRef.get();
     if (exception != null) {
```

However, this by itself lacks the other feature of a modal dialog: namely blocking user input outside of the dialog.  Therefore, the user can continue to interact with the UI while the background task is running.

One way to block user input could be to use a glass pane over the main frame while the background task is running, but this starts really getting into the weeds.

I suppose a possible hack could be to use a transparent or zero-size modal dialog in place of `WaitDialog`.  That might have the same effect without the visual wait indicator.  I tried this out and it seems to have the desired behavior:

```diff
diff --git a/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java b/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java
index 878a31894..6f8693f08 100644
--- a/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java
+++ b/src/main/java/games/strategy/engine/framework/ui/background/BackgroundTaskRunner.java
@@ -8,6 +8,7 @@ import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
 
 import javax.annotation.Nullable;
+import javax.swing.JDialog;
 import javax.swing.SwingUtilities;
 import javax.swing.SwingWorker;
 
@@ -83,7 +84,8 @@ public final class BackgroundTaskRunner {
 
     final AtomicReference<T> resultRef = new AtomicReference<>();
     final AtomicReference<Throwable> exceptionRef = new AtomicReference<>();
-    final WaitDialog waitDialog = GameRunner.newWaitDialog(message);
+    final JDialog waitDialog = new JDialog(GameRunner.mainFrame, true);
+    waitDialog.setUndecorated(true);
     final SwingWorker<T, Void> worker = new SwingWorker<T, Void>() {
       @Override
       protected T doInBackground() throws Exception {
```

(Note that I had to make `GameRunner#mainFrame` public to get this to work.  The real solution would probably be to add a `GameRunner#newModalDialog()` method or something.)

**Comment by RoiEXLab** [comment]
Secondary loop was exactly what I was looking for.
The dialog will continue to persist, it just won't be tied to the newgamechoosermodel anymore.

**Comment by RoiEXLab** [comment]
@ssoloff Using secondary loop, I was able to add some tests using mocks, but I needed to pass a callback, because 2 SecondaryLoops (one from the dialog, one from my implementation) do not work well with each other. (I ran into the problem that the gamechooser dialog did pop up only after the I manually closed the waiting dialog, because the dialog continued to block SecondaryLoop#enter, because I couldn't get WaitingDialog#setVisible execute before SecondaryLoop#enter)
We really need to increase the functionality of BackgroundTaskRunner, but given your code suggestions I'd probably better leave this up to you at some point in the future.

**Comment by RoiEXLab** [comment]
Unrelated to this PR, but we probably want to rename all the NewGameChooser* types to GameChoose* simply because it's annyoing to write new New.
I just did't want to clutter the diff anymore.

File: src/main/java/games/strategy/engine/framework/ui/NewGameChooser.java [hunk 2]
```diff
@@ -193,9 +198,22 @@ public void mouseExited(final MouseEvent e) {}
...
   }
 
-  /** Populates the NewGameChooserModel cache if empty, then returns the cached instance. */
+  /**
+   * Populates the NewGameChooserModel cache if empty, then returns the cached instance.
+   * 
+   * @param doneAction A Runnable being executed on the EDT after the NewGameChooserModel has
+   *        been instantiated.
+   */
+  public static synchronized NewGameChooserModel getNewGameChooserModel(final Runnable doneAction) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/ui/NewGameChooser.java]: It's not clear why this static method (as well as the `getNewGameChooserModel()` overload below) needs to be synchronized.  I don't see any static state in this class that is modified as a result of these calls.

Because of the constraint that a `NewGameChooserModel` must be constructed on the EDT, these methods should never be called from a thread other than the EDT anyway.  So, if only a single thread will ever call these methods, the synchronization isn't necessary.

File: src/main/java/games/strategy/engine/framework/ui/NewGameChooserEntry.java [hunk 3]
```diff
@@ -186,4 +171,14 @@ public boolean equals(final Object obj) {
...
     return this.gameNameAndMapNameProperty.equals(other.gameNameAndMapNameProperty);
   }
+
+
+  @Override
+  public int compareTo(NewGameChooserEntry o) {
+    return getLowerCaseComparable(this).compareTo(getLowerCaseComparable(o));
+  }
+
+  private String getLowerCaseComparable(final NewGameChooserEntry newGameChooserEntry) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/ui/NewGameChooserEntry.java]: After moving this method to `NewGameChooserEntry`, it can now be declared static.

File: src/main/java/games/strategy/engine/framework/ui/NewGameChooserModel.java [hunk 4]
```diff
@@ -38,15 +44,41 @@
...
+    final SecondaryLoop loop = Toolkit.getDefaultToolkit().getSystemEventQueue().createSecondaryLoop();
+
+    new SwingWorker<Void, Void>() {
+      @Override
+      protected Void doInBackground() {
+        final List<NewGameChooserEntry> entries = new ArrayList<>(mapSupplier.get());
+        Collections.sort(entries);
+
+        for (final NewGameChooserEntry entry : entries) {
+          SwingUtilities.invokeLater(() -> addElement(entry));
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/ui/NewGameChooserModel.java]: This is kind of an anti-pattern when using `SwingWorker`.  Idiomatic usage would be to return `entries` as the task return value, and then call `addElement()` on the entries in `done()`.

File: src/test/java/games/strategy/engine/framework/ui/NewGameChooserModelTest.java [hunk 5]
```diff
@@ -1,12 +1,40 @@
...
+    try {
+      SwingUtilities.invokeAndWait(() -> {
+        final Runnable doneAction = mock(Runnable.class);
+        final NewGameChooserEntry entry = mock(NewGameChooserEntry.class);
+        final NewGameChooserModel model = new NewGameChooserModel(doneAction, () -> Collections.singleton(entry));
+        assertEquals(entry, model.get(0));
+        verify(doneAction).run();
+      });
+    } catch (InvocationTargetException e) {
+      throw e.getCause();
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/framework/ui/NewGameChooserModelTest.java]: You can clean up the test a little if you just allow the `InvocationTargetException` to be thrown.  The cause should still be captured in the resulting JUnit error stack trace.

File: src/main/java/games/strategy/engine/framework/ui/NewGameChooserEntry.java (hunk 3 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/framework/ui/NewGameChooserEntry.java]: Or we make use of the current instance we have...

File: src/main/java/games/strategy/engine/framework/ui/NewGameChooser.java (hunk 2 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/framework/ui/NewGameChooser.java]: I think the original intention was to avoid 2 threads constructing 2 model instances at the same time because of the underlying IO operations, but java should really be able to handle 2 threads reading the same file at the same time...

Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/framework/ui/NewGameChooser.java]: And after reading the javadoc, it seems that the model used to cache it's requests, so that the directory isn't parsed every time the user clicks on that button. Removing this method in favour of the normal constructor

Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/framework/ui/NewGameChooser.java]: It seems https://github.com/triplea-game/triplea/commit/a8ee6c81ade5917de357ff83f50bf0e26dfdd63a#diff-c8ac1e7aaa575bf40e562d87db225f85L209 removed this functionality, which made some method calls useless. I'm removing those calls instead of fixing them, simply because for one they work already (and removing them gives a performance boost), and most importantly changing this behavior would require passing instances of GameChooser to retrieve the currently active model.
We should make it a hidden option (for MapMakers), to either cache a single instance for the lifetime of TripleA, or to reload it every time the user clicks on this button, or what would be even better, have a dynamic system which updates the model every time the folder changes. (Using some of the nio APIs), but out-of-scope for this PR

File: src/main/java/games/strategy/engine/framework/startup/ui/GameSelectorPanel.java [hunk 6]
```diff
@@ -370,26 +371,40 @@ private void selectGameFile(final boolean saved) {
...
-            return;
-          }
+        try {
+          BackgroundTaskRunner.runInBackgroundAndReturn("Loading map...", () -> {
+            if (!entry.isGameDataLoaded()) {
+              try {
+                entry.fullyParseGameData();
+              } catch (final GameParseException e) {
+                entry.delayParseGameData();
+                NewGameChooser.getNewGameChooserModel().removeEntry(entry);
```
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/framework/startup/ui/GameSelectorPanel.java]: This method removes an entry from a newly created instance anyways, so I removed it.
Like I explained earlier, this method used to return cached instances

**Comment by RoiEXLab** [comment]
@ssoloff I fixed some of the thread safety issues you mentioned.
I also slightly increased the scope of this PR, to be able to rename classes and fix constructors etc.

File: src/main/java/games/strategy/engine/framework/startup/ui/GameSelectorPanel.java [hunk 7]
```diff
@@ -370,25 +371,41 @@ private void selectGameFile(final boolean saved) {
...
-          try {
-            entry.fullyParseGameData();
-          } catch (final GameParseException e) {
-            entry.delayParseGameData();
-            NewGameChooser.getNewGameChooserModel().removeEntry(entry);
-            return;
-          }
+        try {
+          BackgroundTaskRunner.runInBackgroundAndReturn("Loading map...", () -> {
+            synchronized (entry) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/startup/ui/GameSelectorPanel.java]: Not sure this is doing anything useful in terms of thread safety.  The original concern here was publishing values written by one thread to other threads that read it, not necessarily concurrent access (which doesn't appear possible here).

However, after we discussed in your other PR what `Future#get()` is doing behind the scenes in terms of creating a memory barrier, I'm going to assume `SwingWorker#get()` makes the same guarantee (in fact, it looks like `SwingWorker` uses `Future` internally).  So, I think it's safe to assume that values written by the background thread will be published correctly to the EDT once `runInBackgroundAndReturn()` is complete.

TL;DR: Remove the `synchronized` block; your original code is ok. :smile:

File: src/main/java/games/strategy/engine/framework/ui/GameChooserEntry.java [hunk 8]
```diff
@@ -186,4 +171,14 @@ public boolean equals(final Object obj) {
     }
     return this.gameNameAndMapNameProperty.equals(other.gameNameAndMapNameProperty);
   }
+
+
+  @Override
+  public int compareTo(GameChooserEntry o) {
+    return getLowerCaseComparable().compareTo(o.getLowerCaseComparable());
```
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/framework/ui/GameChooserEntry.java]: Right.

File: src/main/java/games/strategy/engine/framework/ui/GameChooserModel.java [hunk 9]
```diff
@@ -31,26 +40,65 @@
...
 
-    for (final NewGameChooserEntry entry : entries) {
-      addElement(entry);
-    }
+  GameChooserModel(final Runnable doneAction, final Supplier<Set<GameChooserEntry>> mapSupplier) {
+    Preconditions.checkState(SwingUtilities.isEventDispatchThread());
+
+    final SecondaryLoop loop = Toolkit.getDefaultToolkit().getSystemEventQueue().createSecondaryLoop();
+
+    new SwingWorker<Collection<GameChooserEntry>, Void>() {
```
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply src/main/java/games/strategy/engine/framework/ui/GameChooserModel.java]: Your Patch is just right!

**Comment by RoiEXLab** [comment]
@ssoloff Sorry to increase the scope of this PR further, but I had to do similar changes like in #2571, simply because the general structure of the class had the same methods (populateFromZip/Folder parseAllMaps etc.)
This resulted in decreased loading times for the select map dialog by a factor of 5 for me.
Of course the final value can be adjusted

**Comment by RoiEXLab** [comment]
I rebased onto latest master and couldn't notice any issues (This _was_ actually intended by me ^^)
